rdrf/rdrf/reports/generator.py

```diff
@@ -0,0 +1,707 @@
+    def postgres(self):
+        column = alc.Column(self.name,
+                            self.datatype,
+                            nullable=True)
+
+        datasource = DataSource(self.registry_model,
+                                column=column,
+                                form_model=self.form_model,
+                                section_model=self.section_model,
+                                cde_model=self.cde_model)
+
+        self.column_map[column] = datasource
+        return column
+
+
+class ClinicalTable(object):
+    def __init__(self, table_type,
+                 table,
+                 columns,
+                 form_model,
+                 section_model=None):
+        self.table_type = table_type
+        self.table = table
+        self.columns = columns
+        self.form_model = form_model
+        self.section_model = section_model
+
+    @property
+    def is_multisection(self):
+        return self.section_model and self.section_model.allow_multiple
+
+    def __str__(self):
+        return self.table.name
+
+
+class MultiSectionExtractor(object):
+    def __init__(self, registry_model, clinical_table, datasources):
+        self.registry_model = registry_model
+        self.clinical_table = clinical_table
+        self.datasources = datasources
+
+    def get_rows(self, patient_model, context_model):
+
+        nested_clinical_data = get_nested_clinical_data(self.registry_model.code,
+                                                 patient_model.pk,
+                                                 context_model.pk)
+        
+        items_list = patient_model.evaluate_field_expression(self.registry_model,
+                                                             self.field_expression,
+                                                             clinical_data=nested_clinical_data)
+
+        return self._convert_to_rows(patient_model, context_model, items_list)
+
+    def _convert_to_rows(self, patient_model, context_model, items_list):
+        form_timestamp = patient_model.get_form_timestamp(
+            self.clinical_table.form_model, context_model=context_model)
+
+        for index, item_dict in enumerate(items_list):
+            item_number = index + 1
+            row_dict = {}
+            row_dict["form"] = self.clinical_table.form_model.name
+            row_dict["section"] = self.clinical_table.section_model.display_name
+            row_dict["item"] = item_number
+            row_dict["patient_id"] = patient_model.pk
+            row_dict["context_id"] = context_model.pk
+            row_dict["timestamp"] = form_timestamp
+        
+            for cde_code in item_dict:
+                cde_model = CommonDataElement.objects.get(code=cde_code)
+                is_file = cde_model.datatype.strip().lower() == "file"
+                    
+                #column_name = nice_name(cde_model.name.lower().strip())
+                column_name = cde_code.replace(" ", "").lower()
```

Comment:
maybe extract function that removes spaces and lowers a string? It is repeated all over the place.